### ISSUE TITLE
double filenames in MEGA, upload 

### ISSUE THREAD

**Issue by dima72** [meganz/sdk#124]
Dear All,
is there way to upload above existing file and replace its content without deleting the file?
MEGA allows to have double filenames, so when I upload above existing file, the double files are created.

thanks,

**Comment by sergiohs84** [comment]
There are different workarounds:

- Use the synchronization capability of the SDK. By doing this, you can setup a local folder to be synced with a remote folder. The SDK will automatically associate each synced local file to the handle of its corresponding node in MEGA. If a local file is changed, then the previous remote file will be first moved to the Rubbish bin (SyncDebris subfolder) and then the new version will be uploaded.
- Manage the local file and remote node linkage by your own and perform the required actions.

In short, if you upload a file that already exists in the remote path, you need to take care and avoid the creation of duplicates.

Regards

**Comment by dima72** [comment]
yes. We do as you tell in our synchronizator.
but weakest link is:
If synchronization interrupted right after remote file is deleted - then file on remote is disappear and user always thinks that software is inconsistent and causes files lost.

**Comment by javiserrano** [comment]
Sorry for my late reply.

Our sync engine atomically puts uploaded files and moves the old ones to the `SyncDebris` folder in the `Rubbish Bin`. It's possible to do the same using the low-level SDK `MegaClient` but it's not possible to do it using the intermediate layer `MegaApi`.

I think that the most appropriate for this use case would be a new function similar to this one that allows to atomically complete an upload and move another node:
`MegaApi::startUploadAndMove(const char *localPath, MegaNode *destinationFolderForUpload, MegaNode* nodeToMove, MegaNode *destinationFolderForMovedNode)`

Would it be OK for you?

**Comment by dima72** [comment]
I see, to use nodeToMove of SyncDerbis as move destination.  Please remind me how to get SyncDerbis node in MegaApi.

**Comment by javiserrano** [comment]
`nodeToMove` would be the node to replace, and `destinationFolderForMovedNode` would be the destination folder for it as soon as the upload finishes. The completion of the upload and the move would be made at once, so you wouldn't have inconsistencies.

As `destinationFolderForMovedNode` you could use `SyncDebris` or the folder that you prefer. `SyncDebris` is just the folder used by our sync algorithm in the SDK.

To get the SyncDebris folder, you can do: `megaApi->getNodeByPath("//bin/SyncDebris")` but it could not exist. In that case, you would have to create it before:
```
MegaNode *rubbish = megaApi->getRubbishNode();
megaApi->createFolder("SyncDebris", rubbish, listener);
delete rubbish;
```

Anyway, I have been thinking about it and maybe it's not a good idea because a function like that would require effort from app developers, while `MegaApi` is intended to make things as easy as possible for them. Also, it could solve the problem with uploads, but it would persist with remote copies or imported nodes.

Maybe it would be better to have something like this:
- `MegaApi::allowDuplicates(bool)`
- `MegaApi::setFolderForDuplicates(MegaNode)`

That way, app developers could use the first function to prevent the creation of files with the same name in the same folder, they could select the destination folder for replaced files, and the SDK would automatically manage them for all operations, not only uploads.

What do you think?

**Comment by dima72** [comment]
I'm guessing MegaApi::setFolderForDuplicates(MegaNode) in your scenario requires rubbish node, and developers will need to get and create it if not exists. Then MegaApi::startUploadAndMove method would be more transparent.

**Comment by javiserrano** [comment]
Please take a look at this branch:
https://github.com/meganz/sdk/tree/upload-and-move

The last commit adds that function:
https://github.com/meganz/sdk/commit/14d3762853e9712bfeb47b0fd4ccbf54b7a7d37b

If it fits your needs we could review it and merge it in the `master` branch.

**Comment by dima72** [comment]
sir, thank you for startUploadAndMove feature. I will be waiting for UploadFromStream feature and will update our product at once. For now we are in beta test.

**Comment by javiserrano** [comment]
Perfect. I will keep this issue open until this function is tested and merged in `master`.

**Comment by dima72** [comment]
my apologies for the delay,

` void startUploadAndMove(const char* localPath, MegaNode* destinationForUpload, MegaNode* nodeToMove, MegaNode *destinationForMove, MegaTransferListener *listener = NULL);
`
Javi, how to obtain **nodeToMove** if that node is not exists yet on remote?
Can **destinationForMove**  be NULL? 

thanks in advance,

**Comment by dima72** [comment]
In general startUploadAndMove works. As **nodeToMove**  and **destinationForMove** i use the same existing node.
But sometimes all transfer crashes at assertion in transfer.cpp line 483 ( please see attach screenshoot ).
I hope if you merge startUploadAndMove into current trunk all is going to work stable.

![assertion](https://cloud.githubusercontent.com/assets/4633233/22655144/b5c04566-ec3b-11e6-9733-f8d1fc00a7da.jpg)

**Comment by dima72** [comment]
up

**Comment by javiserrano** [comment]
Hello,

I have ported `startUploadAndMove` to the current code on `master`. You have the result in the new branch `upload-and-move2`.

Finally, I think that this function won't be merged into `master` because it's being planned to implement support for file versioning, and that would be the preferred approach to replace files with new ones.

Anyway, when that function is ready, I think that it should fit your needs better than the approach in  `upload-and-move2`.

Please, do not use the same node in `nodeToMove` and `destinationForMove`, as you say in your previous comment, because `destinationForMove` must be a folder. If you use a file node, the node , `nodeToMove`  will be moved to the Rubbish Bin.

Regards, Javi.
